### PR TITLE
 feat: add support for secrets in CDK apps

### LINKED ISSUE #656: Use secure values for SSM Parameter store based secrets
CloudFormation recently [launched](https://aws.amazon.com/about-aws/whats-new/2018/08/aws-cloudformation-introduces-dynamic-references-to-support-aws-) support for secure values, so we should use them for the CDK [SecretParameter](https://awslabs.github.io/aws-cdk/refs/_aws-cdk_cdk.html?highlight=secret#@aws-cdk/cdk.SecretParameter)

### PR DESCRIPTION
**Pull request by rix0rrr**
Adds a construct that evaluates to a Secrets Manager secret value (not
SSM because CloudFormation does not yet support SSMPS SecureStrings).

In addition, a toolkit command has been added which can list all the secrets
used in the application and write values for them to Secrets Manager.

No toolkit integ test yet, since integ tests do an 'npm install' for
construct libraries and the published version of the libraries do not
have the required construct.

Fixes #656.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

**Comment by rix0rrr** [comment]
After talking it over we've decided to get rid of the UI. There's too much risk of us getting it wrong.

**Comment by rix0rrr** [comment]
After talking it over we've decided to get rid of the UI. There's too much risk of us getting it wrong.

File: docs/src/passing-in-data.rst [hunk 1]
```diff
@@ -83,15 +84,66 @@ which gets the value of the environment variable **MYBUCKET**.
...
+========================================
+
+To use values from AWS Secrets Manager in your CDK app, create an instance of :py:class:`SecretsManagerValue
+<@aws-cdk/cdk.SecretsManagerValue>`. It represents a value that is retrieved from Secrets Manager and used
+at CloudFormation deployment time.
 
 .. code-block:: ts
 
-    const myvalue = new SSMParameterProvider(this).getString("my-awesome-value");
+    const loginSecret = new cdk.SecretsManagerString(stack, 'Secret', {
```
Comment (eladb) [inline docs/src/passing-in-data.rst]: Why isn't this class in the aws-secretsmanager module?

File: packages/@aws-cdk/cdk/lib/cloudformation/dynamic-reference.ts [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+export interface SsmParameterStoreStringProps {
+  /**
+   * The name of the parameter store value
+   */
+  parameterName: string;
+
+  /**
+   * The version number of the value you wish to retrieve.
+   */
+  version: number;
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/cloudformation/dynamic-reference.ts]: Is version required?

File: packages/@aws-cdk/cdk/lib/cloudformation/dynamic-reference.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+/**
+ * References a secret value in Secrets Manager
+ *
+ * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-secretsmanager-secret.html
+ */
+export class SecretsManagerString extends DynamicReference {
+  constructor(parent: Construct, id: string, private readonly props: SecretsManagerStringProps) {
+    super(parent, id, {
+      service: DynamicReferenceService.SecretsManager,
+      referenceKey: '',
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/cloudformation/dynamic-reference.ts]: So maybe make this optional in the base class?

File: packages/@aws-cdk/cdk/lib/cloudformation/dynamic-reference.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      service: DynamicReferenceService.Ssm,
+      referenceKey: `${props.parameterName}:${props.version}`,
+    });
+  }
+}
+
+/**
+ * Properties for a ParameterStoreValue
+ */
+export interface SsmParameterStoreSecureStringProps {
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/cloudformation/dynamic-reference.ts]: Intuitively this needs to be in the SSM module and the secrets manager value needs to be in the secrets manager module

File: packages/@aws-cdk/cx-api/lib/metadata/assets.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
+export const ASSET_METADATA = 'aws:cdk:asset';
+
+export interface FileAssetMetadataEntry {
```
Comment (eladb) [inline packages/@aws-cdk/cx-api/lib/metadata/assets.ts]: Maybe split to a separate PR. I believe this is not needed anymore, correct?

File: docs/src/passing-in-data.rst [hunk 6]
```diff
@@ -83,15 +84,66 @@ which gets the value of the environment variable **MYBUCKET**.
...
+Getting a Value from AWS Secrets Manager
+========================================
+
+To use values from AWS Secrets Manager in your CDK app, create an instance of :py:class:`SecretsManagerValue
+<@aws-cdk/cdk.SecretsManagerValue>`. It represents a value that is retrieved from Secrets Manager and used
+at CloudFormation deployment time.
 
 .. code-block:: ts
 
-    const myvalue = new SSMParameterProvider(this).getString("my-awesome-value");
```
Comment (eladb) [inline docs/src/passing-in-data.rst]: We should still document SSMParameterProvider I suppose, no?

File: packages/@aws-cdk/cdk/lib/cloudformation/dynamic-reference.ts (hunk 2 above)
Comment (rix0rrr) [reply packages/@aws-cdk/cdk/lib/cloudformation/dynamic-reference.ts]: Yes

File: docs/src/passing-in-data.rst (hunk 6 above)
Comment (rix0rrr) [reply docs/src/passing-in-data.rst]: It is, the diff is just whack.

File: packages/@aws-cdk/cx-api/lib/metadata/assets.ts (hunk 5 above)
Comment (eladb) [reply packages/@aws-cdk/cx-api/lib/metadata/assets.ts]: I support this general organization but prefer to split the concerns

File: docs/src/passing-in-data.rst (hunk 6 above)
Comment (eladb) [reply docs/src/passing-in-data.rst]: okay!

File: docs/src/passing-in-data.rst [hunk 7]
```diff
@@ -83,15 +84,75 @@ which gets the value of the environment variable **MYBUCKET**.
...
+at CloudFormation deployment time.
+
+.. code-block:: ts
+
+    import secretsmanager = require('@aws-cdk/aws-secretsmanager');
+
+    const loginSecret = new secretsmanager.SecretString(stack, 'Secret', {
+      secretId: 'MyLogin'
+
+      // By default, the latest version is retrieved. It's possible to
```
Comment (RomainMuller) [inline docs/src/passing-in-data.rst]: Optional - I like to use a different comment style for "commented code" and "text", e.g. use `/*` style for text. I find this makes it easier to spot opt-in code fragments.

File: docs/src/passing-in-data.rst [hunk 8]
```diff
@@ -83,15 +84,75 @@ which gets the value of the environment variable **MYBUCKET**.
...
+      // use a specific version instead.
+      // versionStage: 'AWSCURRENT'
+    });
+
+    // Retrieve a value from the secret's JSON
+    const username = loginSecret.jsonValue('username');
+    const password = loginSecret.jsonValue('password');
+
+    // Retrieve the whole secret's string value
+    const fullValue = loginSecret.value;
```
Comment (RomainMuller) [inline docs/src/passing-in-data.rst]: I find it somewhat uncanny that the "field" getters are `jsonValue`, and this is `value` - is this getting a JSON object, or a string containing the JSON object?

File: packages/@aws-cdk/aws-secretsmanager/lib/secret-string.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+ * Properties for a SecretString
+ */
+export interface SecretStringProps {
+  /**
+   * Unique identifier or ARN of the secret
+   */
+  secretId: string;
+
+  /**
+   * Specifies the secret version that you want to retrieve by the staging label attached to the version.
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-secretsmanager/lib/secret-string.ts]: Hopefully, people who need to use this will know what that means...

File: docs/src/passing-in-data.rst (hunk 8 above)
Comment (RomainMuller) [reply docs/src/passing-in-data.rst]: `jsonField(name: string)` may be better, but still #meh. I can't come up with anything that feels substantially better :( Oh well.

**Comment by lestephane** [comment]
This is more aptly named "feat: add support for *reading* secrets in CDK apps". I have yet to find a way to generate secret in a CDK stack. Is that ever going to be supported?
